packages/blocks/src/surface-block/managers/edit-session.ts

```diff
@@ -0,0 +1,318 @@
+    fontWeight: CanvasTextFontWeightSchema,
+    fontStyle: CanvasTextFontStyleSchema,
+    fontSize: z.number(),
+    hasMaxWidth: z.boolean(),
+  }),
+  'affine:note': z.object({
+    background: NoteColorsSchema,
+    hidden: z.boolean(),
+    edgeless: z.object({
+      style: z.object({
+        borderRadius: z.number(),
+        borderSize: z.number(),
+        borderStyle: StorkeStyleSchema,
+        shadowType: NoteShadowsSchema,
+      }),
+    }),
+  }),
+});
+
+export type LastProps = z.infer<typeof LastPropsSchema>;
+
+const SESSION_PROP_KEY = 'blocksuite:prop:record';
+
+const SessionPropsSchema = z.object({
+  viewport: z.union([
+    z.object({
+      centerX: z.number(),
+      centerY: z.number(),
+      zoom: z.number(),
+    }),
+    z.object({
+      referenceId: z.string(),
+      xywh: z.string(),
+      padding: z
+        .tuple([z.number(), z.number(), z.number(), z.number()])
+        .optional(),
+    }),
+  ]),
+  presentBlackBackground: z.boolean(),
+  presentFillScreen: z.boolean(),
+  presentHideToolbar: z.boolean(),
+  templateCache: z.string(),
+  remoteColor: z.string(),
+});
+
+export type SessionProps = z.infer<typeof SessionPropsSchema>;
+
+export type SerializedViewport = z.infer<
+  typeof SessionPropsSchema.shape.viewport
+>;
+
+export class EditSessionStorage {
+  private _lastProps = {
+    connector: {
+      rough: false,
+      frontEndpointStyle: DEFAULT_FRONT_END_POINT_STYLE,
+      rearEndpointStyle: DEFAULT_REAR_END_POINT_STYLE,
+      stroke: GET_DEFAULT_LINE_COLOR(),
+      strokeStyle: StrokeStyle.Solid,
+      strokeWidth: LineWidth.LINE_WIDTH_TWO,
+    },
+    brush: {
+      color: GET_DEFAULT_LINE_COLOR(),
+      lineWidth: LineWidth.Thin,
+    },
+    shape: {
+      shapeType: ShapeType.Rect,
+      fillColor: DEFAULT_SHAPE_FILL_COLOR,
+      strokeColor: DEFAULT_SHAPE_STROKE_COLOR,
+      shapeStyle: ShapeStyle.General,
+    },
+    text: {
+      color: GET_DEFAULT_TEXT_COLOR(),
+      fontFamily: CanvasTextFontFamily.Inter,
+      textAlign: TextAlign.Left,
+      fontWeight: CanvasTextFontWeight.Regular,
+      fontStyle: CanvasTextFontStyle.Normal,
+      fontSize: 24,
+      hasMaxWidth: false,
+    },
+    'affine:note': {
+      background: DEFAULT_NOTE_COLOR,
+      hidden: false,
+      edgeless: {
+        style: {
+          borderRadius: 8,
+          borderSize: 4,
+          borderStyle: StrokeStyle.Solid,
+          shadowType: NOTE_SHADOWS[1],
+        },
+      },
+    },
+  } satisfies LastProps;
+
+  slots = {
+    lastPropsUpdated: new Slot<{
+      type: keyof LastProps;
+      props: Record<string, unknown>;
+    }>(),
+  };
+
+  constructor(private _service: SurfaceService) {
+    this._tryLoadLastProps();
+  }
+
+  private _tryLoadLastProps() {
+    const props = sessionStorage.getItem(SESSION_PROP_KEY);
+    if (!props) return;
+
+    try {
+      const _props = LastPropsSchema.parse(JSON.parse(props));
+      recursive(this._lastProps, _props);
```

Comment:
It's a good step switching to zod. But with this typed schema written, I'm not sure if the `merge` library is appropriate to be used together with it. See [snippet](https://chat.openai.com/share/06ae83df-3b4c-4d28-b498-0299498027fe).